File: semi.ts
type Vector = number[];
type Matrix = Vector[];

type Sprite = {
  kind: string; // ,
  pos: Vector; // ,
};

function matVecMul(mat: Matrix, vec: Vector): Vector {
  // return mat.map((row) =>
  //   row.reduce((sum, value, index) => sum + value * vec[index], 0)
  // );
  let result = new Array(mat.length).fill(0) as Vector;
  for (let i = 0; i < mat.length; i += 1) {
    for (let j = 0; j < vec.length; j += 1) {
      result[i] += mat[i][j] * vec[j];
    }
  }
  return result;
}

function rot(turn: number): Matrix {
  let { PI, cos, sin } = Math;
  let rad = 2 * PI * turn;
  // + 1
  return [[cos(rad), -sin(rad)], [sin(rad), cos(rad)]];
}

function main() {
  let sprite = { kind: "robot", pos: [1, 1] } as Sprite;
  let viewPos = matVecMul(rot(1 / 4), sprite.pos);
  console.log(`${sprite.kind} rotated to ${viewPos}`);
}

main();

File: semi.py
from typing import List, NamedTuple


Vector = List[float]
Matrix = List[Vector]


class Sprite(NamedTuple):
    kind: str  # ;
    pos: Vector


def mat_vec_mul(mat: Matrix, vec: Vector) -> Vector:
    # return [sum(a * b for a, b in zip(row, vec)) for row in mat]
    result = [0.0] * len(mat)
    for i, row in enumerate(mat):
        for (a, b) in zip(row, vec):
            result[i] += a * b
    return result


def rot(turn: float) -> Matrix:
    from math import cos, pi, sin

    rad = \
        2 * pi * turn
    # + 1
    return [[cos(rad), -sin(rad)], [sin(rad), cos(rad)]]


def main():
    sprite = Sprite(kind="robot", pos=[1.0, 1.0])
    view_pos = mat_vec_mul(rot(1 / 4), sprite.pos)
    print(f"{sprite.kind} rotated to {view_pos}")


if __name__ == "__main__":
    main()
